minimint-api/src/module/audit.rs

```diff
@@ -0,0 +1,58 @@
+use crate::db::{Database, DatabaseKeyPrefix, DatabaseKeyPrefixConst};
+
+use std::fmt::Write;
+use std::sync::Arc;
+
+pub struct AuditItem {
+    database_key: Box<dyn DatabaseKeyPrefix>,
+    milli_sat: i64,
+}
+
+impl AuditItem {
+    pub fn from_db<KP, F>(
+        db: &Arc<dyn Database>,
+        key_prefix: &KP,
+        to_milli_sat: F,
+    ) -> Vec<AuditItem>
+    where
+        KP: DatabaseKeyPrefix + DatabaseKeyPrefixConst + 'static,
+        F: Fn(KP::Value) -> i64,
+    {
+        db.find_by_prefix(key_prefix)
+            .map(|res| {
+                let (key, value) = res.expect("DB error");
+                let milli_sat = to_milli_sat(value);
+                let database_key = Box::new(key);
+                AuditItem {
+                    database_key,
+                    milli_sat,
+                }
+            })
+            .collect()
+    }
+
+    pub fn audit_message(items: &Vec<AuditItem>) -> String {
```

Comment:
If the intention is to write it to some stream (e.g. stdout), why not accept a mut formatter reference as argument (just like `Display`)?

Reply:
Yeah I could change this maybe to a visitor pattern.  One call to sum the audit and only generate the String if we need to debug / panic.

Reply:
Another thing I just noticed: `&[T]` is nearly always preferable to `&Vec<T>`